nia/docs 0.3.0 (#10930)
hashicorp/consul: website/content/docs/nia/configuration.mdx

  - `max` - (string: "20s") The maximum period of time to wait after changes are detected before triggering related tasks. If `min` is set, the default period for `max` is 4 times the value of `min`.
- `log_level` - (string: "INFO") The log level to use for Consul-Terraform-Sync logging. This defaults to "INFO". The available log levels are "TRACE", "DEBUG", "INFO", "WARN", and "ERR".
- `port` - (int: 8558) The port for Consul-Terraform-Sync to use to serve API requests.
- `syslog` - Specifies the syslog server for logging.
  - `enabled` - (bool) Enable syslog logging. Specifying other option also enables syslog logging.
  - `facility` - (string: "local0") Name of the syslog facility to log to.
  - `name` - (string: "consul-terraform-sync") Name to use for the daemon process when logging to syslog.
- `working_dir` - (string: "sync-tasks") The base working directory to manage generated artifacts by Consul-Terraform-Sync, including Terraform configuration files for each task. Each task will have a subdirectory used as the working directory, e.g. `./sync-tasks/task-name`. The working directory for tasks can be overridden using the [`task.working_dir`](#working_dir-1) option.

> - `license_path` <EnterpriseAlert inline /> - (string: "") Configures the path to the file containing the license. License must be set in order to use enterprise but can be alternatively set through environment variables `CONSUL_LICENSE` and `CONSUL_LICENSE_PATH` instead. See section on [setting a license](/docs/nia/enterprise/license#setting-the-license) for more details.

Author: You are correct. I'm not sure if you're asking because the language has some ambiguity. Please let me know if there is and happy to fix that.

Looking at consul docs, the file is likely to have a `.hclic` extension. It's no necessary to have that specific extension since we test with `.lic` file extensions. Do you recommend that we mention the expected extension here?

## Consul

The `consul` block is used to configure Consul-Terraform-Sync connection with a Consul agent to perform queries to the Consul Catalog and Consul KV pertaining to task execution.

> -> **Note:** For improved performance of Consul-Terraform-Sync communicating with the local Consul process you can use HTTP/2. To use HTTP/2 the local Consul needs to have [TLS/HTTPS configured](/docs/agent/options) with, at least, the [cert_file](/docs/agent/options#cert_file) and [key_file](/docs/agent/options#key_file) set. Consul-Terraform-Sync's configuration for HTTP/2 requires `tls.enabled = true` and it must have the address set to the HTTPS URL, `address = example.consul.com:8501`. Optionally you'll need to set `tls.verify = false` if using self-signed certificates for Consul. Using HTTP/2 can solve issues hitting Consul's maximum connection limits as well as just generally perform more efficiently when processing lots of blocking queries.

Reviewer: Thanks for the suggestion @trujillo-adam! Spacing out the note does read better, but there's some limitations around that. The call-out block only supports a contiguous paragraph and even sneaking in `<br>`s break the rendering.

![image](https://user-images.githubusercontent.com/6362111/131372026-c7ce9aa7-2bc0-44ee-95aa-ceb613db3ad6.png)

> -> **Note:** For improved performance of Consul-Terraform-Sync communicating with the local Consul process you can use HTTP/2. To use HTTP/2 the local Consul needs to have [TLS/HTTPS configured](/docs/agent/options) with, at least, the [cert_file](/docs/agent/options#cert_file) and [key_file](/docs/agent/options#key_file) set. Consul-Terraform-Sync's configuration for HTTP/2 requires `tls.enabled = true` and it must have the address set to the HTTPS URL, `address = example.consul.com:8501`. Optionally you'll need to set `tls.verify = false` if using self-signed certificates for Consul. Using HTTP/2 can solve issues hitting Consul's maximum connection limits as well as just generally perform more efficiently when processing lots of blocking queries.

Reviewer: So to iterate on this, I'm thinking have the first and second paragraphs within the callout, and then the last sentence moved to the `tls` config option. Would this work for you?

```suggest
-> **Note:** Use HTTP/2 to improve Consul-Terraform-Sync performance when communicating with the local Consul process. [TLS/HTTPS](/docs/agent/options) must be configured for the local Consul with the [cert_file](/docs/agent/options#cert_file) and [key_file](/docs/agent/options#key_file) parameters set. For the Consul-Terraform-Sync configuration, set `tls.enabled = true` and set the `address` parameter to the HTTPS URL, e.g., `address = example.consul.com:8501`. If using self-signed certificates for Consul, you will also need to set `tls.verify = false`. 

...

- `tls` - Configure TLS to use a secure client connection with Consul. Using HTTP/2 can solve issues related to hitting Consul's maximum connection limits, as well as improve efficiency when processing many blocking queries. This option is required for Consul-Terraform-Sync when connecting to a [Consul agent with TLS verification enabled for HTTPS connections](/docs/agent/options#verify_incoming).
```

> -> **Note:** For improved performance of Consul-Terraform-Sync communicating with the local Consul process you can use HTTP/2. To use HTTP/2 the local Consul needs to have [TLS/HTTPS configured](/docs/agent/options) with, at least, the [cert_file](/docs/agent/options#cert_file) and [key_file](/docs/agent/options#key_file) set. Consul-Terraform-Sync's configuration for HTTP/2 requires `tls.enabled = true` and it must have the address set to the HTTPS URL, `address = example.consul.com:8501`. Optionally you'll need to set `tls.verify = false` if using self-signed certificates for Consul. Using HTTP/2 can solve issues hitting Consul's maximum connection limits as well as just generally perform more efficiently when processing lots of blocking queries.

Reviewer 2: Terrific. If the note blocks get too heavy, it's probably a sign that the information should be integrated into the relevant section or even that a new subtopic should be created.

- `source` - (string: required) Source is the location the driver uses to discover the Terraform module used for automation. The source is the module path which can be local or remote on the [Terraform Registry](https://registry.terraform.io/) or private module registry. Read more on [Terraform module source and other supported types here](https://www.terraform.io/docs/modules/sources.html).
  - To use a private module with the [`terraform` driver](#terraform-driver), run the command [`terraform login [hostname]`](https://learn.hashicorp.com/tutorials/terraform/cloud-login) to authenticate the local Terraform CLI prior to starting Consul-Terraform-Sync.
  - To use a private module with the [`terraform_cloud` driver](#terraform-cloud-driver), no extra steps are needed.
    ```hcl
    // local module example: "./terraform-cts-hello"
    source = "<PATH>"

    // public module example: "mkam/hello/cts"
    source = "<NAMESPACE>/<MODULE NAME>/<PROVIDER>"

    // private module example: "my.tfe.hostname.io/my-org/hello/cts"
    source = "<HOSTNAME>/<ORGANIZATION>/<MODULE NAME>/<PROVIDER>"
    ```

> - `variable_files` - (list[string]) A list of paths to files containing variables for the task. For the [Terraform driver](#terraform-driver), these are used as Terraform [variable definition (`.tfvars`) files](https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files) and consists of only variable name assignments. The variable assignments must match the corresponding variable declarations available by the Terraform module for the task. Consul-Terraform-Sync will generate the intermediate variable declarations to pass as arguments from the auto-generated root module to the task's module. Variables are loaded in the same order as they appear in the order of the files. Duplicate variables are overwritten with the later value. _Note: unless specified by the module, configure arguments for Terraform providers using [`terraform_provider` blocks](#terraform-provider)._

Reviewer: The intention is the files themselves that contains Terraform values.

The abstraction here is residue from us originally writing the docs with flexibility not pinned to Terraform verbiage but that has since changed. We are ok with having Terraform specific verbiage threaded in our docs.

Thanks for pointing out how unclear this portion is for us to revisit! So here's another jab at this section

```suggestion
- `variable_files` - (list[string]) Specifies list of paths to [Terraform variable definition files (`.tfvars`)](https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files). The content of these files should consist of only variable name assignments. The variable assignments must match the corresponding variable declarations made available by the Terraform module for the task. 
  - Variables are loaded in the order they appear in the files. Duplicate variables are overwritten with the later value. _Unless specified by the module, configure arguments for Terraform providers using [`terraform_provider` blocks](#terraform-provider)._
    ```hcl
    # example.tfvars
    address_group = "consul-services"
    tags = [
      "consul-terraform-sync",
      "terraform"
    ]
    ```
```

![image](https://user-images.githubusercontent.com/6362111/131385316-34e0096d-bb2f-44d2-b873-8fe7f6c94330.png)

- `terraform_version` - (string) <EnterpriseAlert inline /> The version of Terraform to use for the Terraform Enterprise workspace associated with the task. Defaults to the latest compatible version supported by the organization. This option is only available when used with the [Terraform Cloud driver](#terraform-cloud-driver); otherwise, set the version within the [Terraform driver](#terraform-driver).

### Task Condition

A `task` block can be optionally configured with a `condition` block to set the conditions that should be met in order to execute that particular task. Below are the different types of conditions that Consul-Terraform-Sync supports.

#### Services Condition

> This is the default type of condition when no `condition` block is configured for a task. This condition will trigger the task on either changes in services configured in [`task.services`](#services) or changes in services that match the regular expression configured in `regexp`. The default is triggering on changes to [`task.services`](#services), and [`task.services`](#services) is required if `regexp` is not configured.

Author: Thanks for the improvements here. To clarify

> "The default condition is to trigger on changes to task.services"

I'm more specifically trying to say (in a wordy way) that the default condition's default trigger is to trigger on task.services. What do you think about updating the above sentence to "The default values to trigger on are the services configured in task.services"?

- `version` - (string) The Terraform version to install and run in automation for task execution. If omitted, the driver will install the latest [compatible release of Terraform](/docs/nia/compatibility#terraform). To change versions, remove the existing binary or change the path to install the desired version. Verify that the desired Terraform version is compatible across all Terraform modules used for Consul-Terraform-Sync automation.
- `working_dir` - (string: "sync-tasks") **Deprecated in Consul-Terraform-Sync 0.3.0**, use the global [`working_dir`](#working_dir) option instead. The base working directory to manage Terraform configurations for all tasks. The full path of each working directory will have the task identifier appended to the end of the path, e.g. `./sync-tasks/task-name`.

## Terraform Cloud Driver

> 

Author: Yes, there is a CTS enterprise binary. CTS OSS does not require a license

<EnterpriseAlert>
  This feature requires{' '}
  <a href="https://www.hashicorp.com/products/consul/features">
    Consul-Terraform-Sync Enterprise
  </a>{' '}
  which is available with <strong>Consul Enterprise</strong>.
</EnterpriseAlert>

The Terraform Cloud driver enables Consul-Terraform-Sync Enterprise to integrate with [**Terraform Enterprise**](https://www.hashicorp.com/products/terraform/editions/enterprise)<sup>1</sup>, the self-hosted distribution of Terraform Cloud. With this driver, Consul-Terraform-Sync automates Terraform runs and remote operations for workspaces.

An overview of features enabled with Terraform Enterprise can be viewed within the [Network Drivers](/docs/nia/network-drivers) documentation.

Only one network driver can be configured per deployment of Consul-Terraform-Sync.

> -> <sup>1</sup> **Upcoming**: The Terraform Cloud driver is introduced in Consul-Terraform-Sync v0.3.0 and only supports [Terraform Enterprise](https://www.hashicorp.com/products/terraform/editions/enterprise), the self-hosted distribution. The upcoming v0.4.0 release will support integration with [Terraform Cloud](https://www.terraform.io/cloud), the HashiCorp managed service.

Reviewer: I agree that we should not include future documentation. We do consider this a special case since it's not common for TFE to release before TFC these days, and are typically released at the same time or TFC first then TFE soon after.

The decision for TFE first was a difficult one for us to work-around and so this callout was added in attempt to address 2 concerns we had:
1. To help with expectations due to the norm of release cadence for TFC and TFE  features
2. The driver is named `terraform_cloud` but only supports TFE would be a non-intuitive user experience.

Does this added context help with why we have these callouts?

  // Optionally set the token to be securely queried from Vault instead of
  // written directly to the configuration file.
  // token = "{{ with secret \"secret/my/path\" }}{{ .Data.data.foo }}{{ end }}"

  required_providers {
    myprovider = {
      source = "namespace/myprovider"
      version = "1.3.0"
    }
  }
}
```

- `hostname` - (string) The Terraform Enterprise hostname to connect to. Can be overridden with the `TFC_HOSTNAME` environment variable.
- `organization` - (string) The Terraform Enterprise organization that hosts the managed workspaces by Consul-Terraform-Sync. Can be overridden with the `TFC_ORGANIZATION` environment variable.
- `token` - (string) Required [Team API token](https://www.terraform.io/docs/cloud/users-teams-organizations/api-tokens.html#team-api-tokens) used for authentication with Terraform Enterprise and workspace management. Only workspace permissions are needed for Consul-Terraform-Sync. The token can also be provided using the `TFC_TOKEN` environment variable.
  - We recommend creating a dedicated team and team API token to isolate automation by Consul-Terraform-Sync from other Terraform Enterprise operations.

> - `workspace_prefix` - (string) A prefix which is prepended to the automatically generated workspace names used for automation. This prefix will be used by all tasks which use this driver. If this prefix is not included, then the workspace will default to the task name.

Author: To clarify, the default behavior (when workspace_prefix is not configured) is for the workspace name to be the same as the task name (no prefix). When workspace_prefix is configured then the workspace name would be the prefix + the task name.

What are your thoughts on updating this to:

>- `workspace_prefix` - (string) Specifies a prefix to prepend to the automatically-generated workspace names used for automation. By default, when `workspace_prefix` is unconfigured, the automatically-generated workspace name is the task name. When `workspace_prefix` is configured, the workspace name will be `<workspace_prefix value><task name>`. For example, if you configure the prefix as "cts_", a task with the name "task_firewall" will have the workspace name of "cts_task_firewall"

> - `required_providers` - (obj: required) Declare each Terraform provider used across all tasks. This is similar to the [Terraform `terraform.required_providers`](https://www.terraform.io/docs/configuration/provider-requirements.html#requiring-providers) field to specify the source and version for each provider. Consul-Terraform-Sync will process these requirements when preparing each task that uses the provider.

Author: Thanks for the question. The Terraform `required_providers` block doesn't have any set parameters dictated by Terraform in the sense that users can configure their providers here based on their provider's configuration parameters.

Do you have any thoughts on replacing "This is similar to the Terraform `terraform.required_providers` field to specify the source and version for each provider."  => "This can be configured the same as how you would configure the Terraform `terraform.required_providers` field to specify the source and version for each provider."
